Source/Internal/IGListPreprocessor.m

```diff
@@ -0,0 +1,98 @@
+//
+
+#import "IGListPreprocessor.h"
+
+#import "IGListAssert.h"
+#import "IGListPreprocessingContext.h"
+#import "IGListPreprocessingDelegate.h"
+#import "IGListSectionController.h"
+#import "IGListSectionMap.h"
+
+@implementation IGListPreprocessor {
+	NSMapTable<IGListPreprocessingContext *, id<IGListPreprocessingDelegate>> *_contextToDelegateMap;
+	IGListSectionMap *_sectionMap;
+	dispatch_group_t _group;
+	dispatch_block_t _completion;
+	BOOL _ranCompletion;
+}
+
+- (instancetype)initWithSectionMap:(IGListSectionMap *)sectionMap {
+	if (self = [super init]) {
+		_contextToDelegateMap = [NSMapTable strongToStrongObjectsMapTable];
+		_sectionMap = [sectionMap copy];
+		_group = dispatch_group_create();
+	}
+	return self;
+}
+
+- (void)schedulePreprocessingWithContainerSize:(CGSize)containerSize completion:(nonnull dispatch_block_t)completionBlock {
+	IGAssertMainThread();
+	_completion = completionBlock;
+	
+	dispatch_queue_t queue = dispatch_queue_create("IGListPreprocessor Queue", DISPATCH_QUEUE_CONCURRENT);
+	
+	dispatch_group_t group = _group;
+	NSMapTable *contextToDelegateMap = _contextToDelegateMap;
+	[_sectionMap enumerateUsingBlock:^(id  _Nonnull object, IGListSectionController<IGListSectionType> * _Nonnull sectionController, NSInteger section, BOOL * _Nonnull stop) {
+		id<IGListPreprocessingDelegate> delegate = sectionController.preprocessingDelegate;
+		if (delegate == nil) {
+			return;
+		}
+		
+		// Create a context and associate it with this delegate.
+		IGListPreprocessingContext *context = [[IGListPreprocessingContext alloc] initWithObject:object containerSize:containerSize sectionIndex:section dispatchGroup:group];
+		[contextToDelegateMap setObject:delegate forKey:context];
+		
+		// Schedule the delegate's work – the context will call preprocessingDidCompleteWithContext:
+		dispatch_group_enter(group);
+		dispatch_async(queue, ^{
+			[delegate preprocessWithContext:context];
+		});
+	}];
+	
+	// When the work is all done, call -runCompletionIfNeeded on main.
+	dispatch_group_notify(_group, dispatch_get_main_queue(), ^{
+		[self runCompletionIfNeeded];
+	});
+}
+
+- (void)waitForPreprocessingToFinish {
+	IGAssertMainThread();
+	
+	dispatch_group_wait(_group, DISPATCH_TIME_FOREVER);
+	[self runCompletionIfNeeded];
+}
+
+#pragma mark - Private
+
+- (void)runCompletionIfNeeded {
+	IGAssertMainThread();
+	
+	// Handle case where we ran completion before e.g.
+	// -waitForPreprocessingToFinish was called before
+	// completion was run naturally.
+	if (_ranCompletion) {
+		return;
+	}
+	_ranCompletion = YES;
+	
+	// Inform all the delegates about preprocessing being done.
+	for (IGListPreprocessingContext *context in _contextToDelegateMap) {
+		id<IGListPreprocessingDelegate> delegate = [_contextToDelegateMap objectForKey:context];
+		[delegate preprocessingDidFinishWithContext:context];
+	}
+	
+	// Call and release our completion block.
+	if (_completion) {
+		_completion();
+		_completion = nil;
+	}
+}
+
+@end
```

Comment:
This is how you write beautiful async code. Note that there are no gross locks, atomic stuff, or custom containers. Just perfectly designed GCD. Kudos @Adlai-Holler ❤️